File: Day45/Part9/battle_royale.py
import math

def dist(p1, p2):
    return math.sqrt((p1[0]-p2[0])**2 + (p1[1]-p2[1])**2)

def solve():
    # Hardcoded Input
    # 3 Players
    # P0: (0,0), HP 50, Atk 10, Rng 2, Spd 1
    # P1: (5,0), HP 50, Atk 10, Rng 2, Spd 1
    # P2: (2,5), HP 60, Atk 5, Rng 5, Spd 0 (Sniper)
    
    players = [
        {'id': 0, 'hp': 50, 'atk': 10, 'rng': 2, 'spd': 1, 'pos': [0.0, 0.0], 'alive': True},
        {'id': 1, 'hp': 50, 'atk': 10, 'rng': 2, 'spd': 1, 'pos': [5.0, 0.0], 'alive': True},
        {'id': 2, 'hp': 60, 'atk': 5,  'rng': 5, 'spd': 0, 'pos': [2.5, 5.0], 'alive': True}
    ]
    
    # Seed Cycle (Permutations of 0, 1, 2)
    # Let's say cycle length 2
    cycle = [
        [0, 1, 2],
        [2, 1, 0]
    ]
    
    turn = 0
    max_turns = 100
    
    while turn < max_turns:
        # Check winner
        alive_ids = [p['id'] for p in players if p['alive']]
        if len(alive_ids) <= 1:
            return alive_ids[0] if alive_ids else -1
            
        # 1. Move
        # All move simultaneously towards nearest enemy
        new_positions = []
        for p in players:
            if not p['alive']:
                new_positions.append(p['pos'])
                continue
                
            # Find nearest enemy
            nearest = None
            min_d = float('inf')
            for other in players:
                if p['id'] != other['id'] and other['alive']:
                    d = dist(p['pos'], other['pos'])
                    if d < min_d:
                        min_d = d
                        nearest = other
            
            if nearest:
                # Move towards nearest
                # Direction
                dx = nearest['pos'][0] - p['pos'][0]
                dy = nearest['pos'][1] - p['pos'][1]
                length = math.sqrt(dx*dx + dy*dy)
                
                if length > 0:
                    move_dist = min(p['spd'], length) # Don't overshoot? Or stop at range?
                    # Simple: Move full speed
                    # But if already in range?
                    # "Bergerak mendekati pemain terdekat. Jika dalam range, mereka menyerang."
                    # Usually move then attack.
                    # If in range, maybe stop?
                    # Let's assume they move to minimize distance.
                    
                    # If dist > range, move.
                    # If dist <= range, maybe stay?
                    # Let's assume aggressive: Always move closer until collision (dist=0).
                    
                    nx = p['pos'][0] + (dx/length) * move_dist
                    ny = p['pos'][1] + (dy/length) * move_dist
                    new_positions.append([nx, ny])
                else:
                    new_positions.append(p['pos'])
            else:
                new_positions.append(p['pos'])
        
        # Update positions
        for i, pos in enumerate(new_positions):
            players[i]['pos'] = pos
            
        # 2. Attack
        # Get permutation
        perm = cycle[turn % len(cycle)]
        
        for pid in perm:
            attacker = players[pid]
            if not attacker['alive']: continue
            
            # Find target in range
            # Nearest in range
            target = None
            min_d = float('inf')
            
            for other in players:
                if attacker['id'] != other['id'] and other['alive']:
                    d = dist(attacker['pos'], other['pos'])
                    if d <= attacker['rng']:
                        if d < min_d:
                            min_d = d
                            target = other
            
            if target:
                target['hp'] -= attacker['atk']
                if target['hp'] <= 0:
                    target['alive'] = False
                    
        turn += 1
        
    return -1 # Timeout

if __name__ == "__main__":
    result = solve()
    print(f"Survivor ID: {result}")

File: learnLeetCode/Day45/Part9/battle_royale.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>
#include <limits>

using namespace std;

struct Player {
    int id;
    double hp;
    double atk;
    double rng;
    double spd;
    double x, y;
    bool alive;
};

double dist(const Player& a, const Player& b) {
    return sqrt(pow(a.x - b.x, 2) + pow(a.y - b.y, 2));
}

int solve() {
    vector<Player> players = {
        {0, 50, 10, 2, 1, 0.0, 0.0, true},
        {1, 50, 10, 2, 1, 5.0, 0.0, true},
        {2, 60, 5, 5, 0, 2.5, 5.0, true}
    };
    
    vector<vector<int>> cycle = {
        {0, 1, 2},
        {2, 1, 0}
    };
    
    int turn = 0;
    int max_turns = 100;
    
    while(turn < max_turns) {
        int alive_count = 0;
        int survivor = -1;
        for(const auto& p : players) if(p.alive) { alive_count++; survivor = p.id; }
        
        if(alive_count <= 1) return survivor;
        
        // 1. Move
        vector<pair<double, double>> new_pos(players.size());
        for(int i=0; i<players.size(); ++i) {
            if(!players[i].alive) {
                new_pos[i] = {players[i].x, players[i].y};
                continue;
            }
            
            int nearest_idx = -1;
            double min_d = 1e9;
            
            for(int j=0; j<players.size(); ++j) {
                if(i != j && players[j].alive) {
                    double d = dist(players[i], players[j]);
                    if(d < min_d) {
                        min_d = d;
                        nearest_idx = j;
                    }
                }
            }
            
            if(nearest_idx != -1) {
                double dx = players[nearest_idx].x - players[i].x;
                double dy = players[nearest_idx].y - players[i].y;
                double len = sqrt(dx*dx + dy*dy);
                
                if(len > 0) {
                    double move = min(players[i].spd, len);
                    new_pos[i] = {players[i].x + (dx/len)*move, players[i].y + (dy/len)*move};
                } else {
                    new_pos[i] = {players[i].x, players[i].y};
                }
            } else {
                new_pos[i] = {players[i].x, players[i].y};
            }
        }
        
        for(int i=0; i<players.size(); ++i) {
            players[i].x = new_pos[i].first;
            players[i].y = new_pos[i].second;
        }
        
        // 2. Attack
        const vector<int>& perm = cycle[turn % cycle.size()];
        for(int pid : perm) {
            if(!players[pid].alive) continue;
            
            int target_idx = -1;
            double min_d = 1e9;
            
            for(int j=0; j<players.size(); ++j) {
                if(pid != j && players[j].alive) {
                    double d = dist(players[pid], players[j]);
                    if(d <= players[pid].rng) {
                        if(d < min_d) {
                            min_d = d;
                            target_idx = j;
                        }
                    }
                }
            }
            
            if(target_idx != -1) {
                players[target_idx].hp -= players[pid].atk;
                if(players[target_idx].hp <= 0) {
                    players[target_idx].alive = false;
                }
            }
        }
        
        turn++;
    }
    
    return -1;
}

int main() {
    cout << "Survivor ID: " << solve() << endl;
    return 0;
}
